segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,394 @@
+                config.getRoot() + streamSegmentName);
+
+        AccessControlList acls = client.getObjectAcl(config.getBucket(), config.getRoot() + streamSegmentName);
+
+        boolean canWrite = acls.getGrants().stream().filter((grant) -> {
+            return grant.getPermission().compareTo(Permission.WRITE) >= 0;
+        }).count() > 0;
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        GetObjectResult<InputStream> result = null;
+        try {
+            result = client.getObject(config.getBucket(), config.getRoot() + streamSegmentName);
+        } catch (S3Exception e) {
+            if ( e.getErrorCode().equals("NoSuchKey")) {
+                result = null;
+            } else {
+                throw e;
+            }
+        }
+        return result != null;
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) throws StreamSegmentExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+
+        if (!client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName)
+                  .getObjects().isEmpty()) {
+            throw new StreamSegmentExistsException(streamSegmentName);
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getBucket(),
+                config.getRoot() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
+
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException, BadOffsetException {
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        if (si.getLength() < offset) {
```

Comment:
@andreipaduroiu is right, there is some mismatch here. The write contract in `Storage` says:

```
/**
     * Writes the given data to the StreamSegment.
     *
     * @param handle  A read-write SegmentHandle that points to a Segment to write to.
     * @param offset  The offset in the StreamSegment to write data at.
     * @param data    An InputStream representing the data to write.
     * @param length  The length of the InputStream.
     * @param timeout Timeout for the operation.
     * @return A CompletableFuture that, when completed, will indicate the operation succeeded. If the operation failed,
     * it will contain the cause of the failure. Notable exceptions:
     * <ul>
     * <li> BadOffsetException: When the given offset does not match the actual length of the segment in storage.
     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
     * <li> StorageNotPrimaryException: When this Storage instance is no longer primary for this Segment (it was fenced out).
     * </ul>
     * @throws IllegalArgumentException If handle is read-only.
     */
```

This implementation doesn't seem to be respecting the condition to throw `BadOffsetException`. If contract needs to be different, then we need to change it accordingly, but we can simply ignore the contract of the interface.